File: kd_tree.cpp
#ifndef LIDAR_LOCALIZATION_TOOLS_KD_TREE_
#define LIDAR_LOCALIZATION_TOOLS_KD_TREE_

#include <Eigen/Dense>
#include <iostream>
#include <memory>
#include <queue>
#include <stack>
#include <vector>

#include "math.h"

namespace lidar_localization {

template <class T>
class KDNode {
 public:
  typedef std::shared_ptr<KDNode<T>> KDNodePtr;
  KDNode() = default;
  ~KDNode() = default;

  void SetParent(const KDNodePtr node_ptr) { parent_ = node_ptr; }
  void SetLeftChild(const KDNodePtr node_ptr) { left_child_ = node_ptr; }
  void SetRightChild(const KDNodePtr node_ptr) { right_child_ = node_ptr; }
  void SetData(const T& data) { data_ = data; }
  void SetSplit(const int split) { split_ = split; }
  void SetDist(const double dist) { dist_ = dist; }

  bool HasParent() const { return parent_ != nullptr; }
  bool HasLeftChild() const { return left_child_ != nullptr; }
  bool HasRightChild() const { return right_child_ != nullptr; }

  bool IsLeftChild(const KDNodePtr node_ptr) { return node_ptr->GetData()[split_] < data_[split_]; }

  KDNodePtr GetParent() const { return parent_; }
  KDNodePtr GetLeftChild() const { return left_child_; }
  KDNodePtr GetRightChild() const { return right_child_; }
  int GetSplit() const { return split_; }
  T GetData() const { return data_; }
  double GetDist() const { return dist_; }

 private:
  int split_ = 0;
  KDNodePtr parent_ = nullptr;
  KDNodePtr left_child_ = nullptr;
  KDNodePtr right_child_ = nullptr;
  T data_;
  double dist_;
};

template <class T>
class KDTree {
 public:
  typedef std::shared_ptr<KDNode<T>> KDNodePtr;

  KDTree() = default;
  ~KDTree() = default;

  void AddData(const T& element, const int dimension) {
    elements_.emplace_back(element);
    dimension_ = dimension;
  }
  void ClearData() { elements_.clear(); }
  bool HasData() const { return !elements_.empty(); }
  size_t DataSize() const { return elements_.size(); }

  bool GenerateTree() {
    if (!HasData() || dimension_ <= 0) {
      return false;
    }
    root_ = CreateNode(0, DataSize() - 1);
    return root_ != nullptr;
  }

  bool TreeValid() const { return root_ != nullptr; }

  KDNodePtr CreateNode(const size_t lower_idx, const size_t upper_idx) {
    KDNodePtr node_ptr(new KDNode<T>);
    if (lower_idx == upper_idx) {
      node_ptr->SetData(elements_[lower_idx]);
      std::cout << "lower_idx: " << lower_idx << " upper_idx: " << upper_idx << std::endl;
      PrintElement(node_ptr->GetData());
      return node_ptr;
    }
    // determine split with largest variation
    int split = 0;
    double max_var = 0.0;
    for (size_t i = 0; i < dimension_; ++i) {
      double var = CalculateVariation(i, lower_idx, upper_idx);
      if (var >= max_var) {
        max_var = var;
        split = i;
      }
    }
    std::cout << "max_var: " << max_var << " split: " << split << std::endl;
    node_ptr->SetSplit(split);
    // sort data set by split
    // and set mid value as node data
    InsertSort(split, lower_idx, upper_idx);
    size_t mid_idx = (lower_idx + upper_idx + 1ul) / 2ul;
    node_ptr->SetData(elements_[mid_idx]);
    std::cout << "lower_idx: " << lower_idx << " upper_idx: " << upper_idx << " mid_idx: " << mid_idx << std::endl;
    PrintElement(node_ptr->GetData());
    // get left & right node
    if (mid_idx > lower_idx) {
      KDNodePtr left_node_ptr = CreateNode(lower_idx, mid_idx - 1);
      left_node_ptr->SetParent(node_ptr);
      node_ptr->SetLeftChild(left_node_ptr);
    }
    if (mid_idx < upper_idx) {
      KDNodePtr right_node_ptr = CreateNode(mid_idx + 1, upper_idx);
      right_node_ptr->SetParent(node_ptr);
      node_ptr->SetRightChild(right_node_ptr);
    }

    return node_ptr;
  }

  bool FindNearestPoint(const T& target_point, T& nearest_point) {
    if (!HasData() || !TreeValid()) {
      return false;
    }

    auto func_is_leaf = [](KDNodePtr node_ptr) { return !(node_ptr->HasLeftChild() || node_ptr->HasRightChild()); };

    auto func_cal_dist_sqr = [this](KDNodePtr node_ptr0, KDNodePtr node_ptr1) {
      double sqr_sum = 0.0;
      for (int i = 0; i < this->dimension_; ++i) {
        sqr_sum += std::pow(node_ptr0->GetData()[i] - node_ptr1->GetData()[i], 2);
      }
      return sqr_sum;
    };

    auto func_judge_hypersurface_cut = [](KDNodePtr current_node_ptr, KDNodePtr target_node_ptr,
                                          const double radius_sqr) {
      auto current_split = current_node_ptr->GetSplit();
      double dist = std::fabs(current_node_ptr->GetData()[current_split] - target_node_ptr->GetData()[current_split]);
      // std::cout << "cur split: " << current_split << " cut dist: " << dist << " radius_sqr: " << radius_sqr <<
      // std::endl;
      return dist * dist <= radius_sqr;
    };

    // binary search & record search path
    KDNodePtr current_node_ptr = root_;
    std::stack<KDNodePtr> node_stack;
    node_stack.push(current_node_ptr);
    while (!func_is_leaf(current_node_ptr)) {
      auto current_node_data = current_node_ptr->GetData();
      auto current_split = current_node_ptr->GetSplit();
      if (current_node_data[current_split] > target_point[current_split] && current_node_ptr->HasLeftChild()) {
        current_node_ptr = current_node_ptr->GetLeftChild();
        node_stack.push(current_node_ptr);
      } else if (current_node_ptr->HasRightChild()) {
        current_node_ptr = current_node_ptr->GetRightChild();
        node_stack.push(current_node_ptr);
      }
    }

    // reverse search to update possible closer node
    KDNodePtr target_node_ptr(new KDNode<T>);
    target_node_ptr->SetData(target_point);
    KDNodePtr nearest_node_ptr = node_stack.top();
    // std::cout << "nearest node 0: ";
    // PrintElement(nearest_node_ptr->GetData());
    node_stack.pop();
    double min_dist_sqr = func_cal_dist_sqr(target_node_ptr, nearest_node_ptr);
    // std::cout << "min_dist_sqr 0: " << min_dist_sqr << std::endl;
    while (!node_stack.empty()) {
      current_node_ptr = node_stack.top();
      node_stack.pop();
      double cur_dist_sqr = func_cal_dist_sqr(target_node_ptr, current_node_ptr);
      // std::cout << "current node: ";
      // PrintElement(current_node_ptr->GetData());
      // std::cout << "cur_dist_sqr: " << cur_dist_sqr << std::endl;
      if (cur_dist_sqr < min_dist_sqr) {
        nearest_node_ptr = current_node_ptr;
        min_dist_sqr = cur_dist_sqr;
        // std::cout << "min_dist_sqr: " << min_dist_sqr << std::endl;
      }
      // std::cout << "current is leaf: " << func_is_leaf(current_node_ptr) << std::endl;
      if (!func_is_leaf(current_node_ptr)) {
        // if node doesn't intersect with target node's hyper surface
        // reverse search is done(if it does, there's a chance the other side
        // has closer dist node)
        if (func_judge_hypersurface_cut(current_node_ptr, target_node_ptr, min_dist_sqr)) {
          // std::cout << "cut with ";
          // PrintElement(current_node_ptr->GetData());
          if (current_node_ptr->IsLeftChild(target_node_ptr) && current_node_ptr->HasRightChild()) {
            current_node_ptr = current_node_ptr->GetRightChild();
            node_stack.push(current_node_ptr);
            // std::cout << "go to right child: ";
            // PrintElement(current_node_ptr->GetData());
          } else if (current_node_ptr->HasLeftChild()) {
            current_node_ptr = current_node_ptr->GetLeftChild();
            node_stack.push(current_node_ptr);
            // std::cout << "go to left child: ";
            // PrintElement(current_node_ptr->GetData());
          }
        }
      }
    }

    nearest_point = nearest_node_ptr->GetData();
    return true;
  }

  bool FindNearestPoint(const T& target_point, std::vector<T>& nearest_point_vector, const size_t k = 1) {
    if (!HasData() || !TreeValid()) {
      return false;
    }

    // recursive lambda function with no return
    auto func_dfs = [this](std::deque<KDNodePtr>& node_pq, KDNodePtr current_node_ptr, KDNodePtr target_node_ptr,
                           auto&& func_dfs) -> void {
      // struct {
      //   bool operator()(const KDNodePtr& a, const KDNodePtr& b) { return a->GetDist() <= b->GetDist(); }
      // } cmp;

      auto func_is_leaf = [](KDNodePtr node_ptr) { return !(node_ptr->HasLeftChild() || node_ptr->HasRightChild()); };

      auto func_cal_dist_sqr = [this](KDNodePtr node_ptr0, KDNodePtr node_ptr1) {
        double sqr_sum = 0.0;
        for (int i = 0; i < this->dimension_; ++i) {
          sqr_sum += std::pow(node_ptr0->GetData()[i] - node_ptr1->GetData()[i], 2);
        }
        return sqr_sum;
      };

      auto func_judge_hypersurface_cut = [](KDNodePtr current_node_ptr, KDNodePtr target_node_ptr,
                                            const double radius_sqr) {
        auto current_split = current_node_ptr->GetSplit();
        double dist = std::fabs(current_node_ptr->GetData()[current_split] - target_node_ptr->GetData()[current_split]);
        std::cout << "cur split: " << current_split << " cut dist: " << dist << " radius_sqr: " << radius_sqr
                  << std::endl;
        return dist * dist <= radius_sqr;
      };

      if (current_node_ptr == nullptr) {
        std::cout << "null & return" << std::endl;
        return;
      }
      std::cout << "current node: ";
      PrintElement(current_node_ptr->GetData());

      // calculate sqr dist and push into pq
      double min_dist_sqr = func_cal_dist_sqr(current_node_ptr, target_node_ptr);
      std::cout << "min_dist_sqr: " << min_dist_sqr << std::endl;
      current_node_ptr->SetDist(min_dist_sqr);
      node_pq.emplace_back(current_node_ptr);

      if (node_pq.size() > this->k_) {
        auto cmp = [](const KDNodePtr& a, const KDNodePtr& b) { return a->GetDist() <= b->GetDist(); };
        std::sort(node_pq.begin(), node_pq.end(), cmp);
        node_pq.pop_back();
      }

      // recursively search the half of the tree
      const auto current_node_data = current_node_ptr->GetData();
      const auto current_split = current_node_ptr->GetSplit();
      const auto target_node_data = target_node_ptr->GetData();
      bool search_direction = 0;
      if (current_node_data[current_split] > target_node_data[current_split]) {
        std::cout << "go to left child: \n";
        func_dfs(node_pq, current_node_ptr->GetLeftChild(), target_node_ptr, func_dfs);
      } else {
        std::cout << "go to right child: \n";
        func_dfs(node_pq, current_node_ptr->GetRightChild(), target_node_ptr, func_dfs);
        search_direction = 1;
      }
      std::cout << "top node: ";
      PrintElement(node_pq.front()->GetData());
      std::cout << "cur node: ";
      PrintElement(current_node_ptr->GetData());
      // If the candidate hypersphere crosses this splitting plane, look on the other side of the plane by examining the
      // other subtree.
      bool size_not_enough = node_pq.size() < this->k_;
      bool cut_with_hypersurface =
          func_judge_hypersurface_cut(current_node_ptr, target_node_ptr, node_pq.front()->GetDist());
      std::cout << "size: " << node_pq.size() << " cut: " << cut_with_hypersurface << std::endl;
      if (size_not_enough || cut_with_hypersurface) {
        if (search_direction == 0) {
          std::cout << "cut & go to right child: ";
          PrintElement(current_node_ptr->GetData());
          func_dfs(node_pq, current_node_ptr->GetRightChild(), target_node_ptr, func_dfs);
        } else {
          std::cout << "cut & go to left child: ";
          PrintElement(current_node_ptr->GetData());
          func_dfs(node_pq, current_node_ptr->GetLeftChild(), target_node_ptr, func_dfs);
        }
      }
      std::cout << "final cur node: ";
      PrintElement(current_node_ptr->GetData());
    };

    k_ = k;
    std::deque<KDNodePtr> node_pq;
    KDNodePtr target_node_ptr(new KDNode<T>);
    target_node_ptr->SetData(target_point);
    // recursive search
    func_dfs(node_pq, root_, target_node_ptr, func_dfs);

    // output to a vector
    nearest_point_vector.clear();
    nearest_point_vector.reserve(k);
    for (const auto& node_ptr : node_pq) {
      nearest_point_vector.emplace_back(node_ptr->GetData());
    }
    return true;
  }

  void PrintTree(const int method = 0) {
    if (!HasData() || !TreeValid()) {
      return;
    }
    if (method == 0) {  // bfs
      std::cout << "[PrintTree] bfs" << std::endl;
      std::queue<KDNodePtr> node_queue;
      node_queue.push(root_);
      while (!node_queue.empty()) {
        KDNodePtr node_ptr = node_queue.front();
        node_queue.pop();
        PrintElement(node_ptr->GetData());
        if (node_ptr->HasLeftChild()) {
          node_queue.push(node_ptr->GetLeftChild());
        }
        if (node_ptr->HasRightChild()) {
          node_queue.push(node_ptr->GetRightChild());
        }
      }
    } else {  // dfs
      // recursive lambda function with no return
      auto func_dfs = [this](std::stack<KDNodePtr>& node_stack, KDNodePtr cur_node_ptr, auto&& func_dfs) -> void {
        int counter = 0;
        if (cur_node_ptr->HasLeftChild()) {
          node_stack.push(cur_node_ptr->GetLeftChild());
          func_dfs(node_stack, cur_node_ptr->GetLeftChild(), func_dfs);
        }
        ++counter;
        if (cur_node_ptr->HasRightChild()) {
          node_stack.push(cur_node_ptr->GetRightChild());
          func_dfs(node_stack, cur_node_ptr->GetRightChild(), func_dfs);
        }
        ++counter;
        if (!(cur_node_ptr->HasLeftChild() || cur_node_ptr->HasRightChild()) || counter == 2) {
          this->PrintElement(cur_node_ptr->GetData());
          node_stack.pop();
        }
      };

      std::cout << "[PrintTree] dfs" << std::endl;
      std::stack<KDNodePtr> node_stack;
      node_stack.push(root_);
      func_dfs(node_stack, root_, func_dfs);
    }
    std::cout << std::endl;
  }

  // from small to big
  void InsertSort(int cur_dim, const size_t lower_idx, const size_t upper_idx) {
    if (!HasData() || lower_idx >= upper_idx || upper_idx >= DataSize()) {
      return;
    }
    for (size_t i = lower_idx + 1; i <= upper_idx; ++i) {
      auto key = elements_[i];
      size_t j = i - 1;
      bool shifted = false;
      bool zero_reached = false;
      while (key[cur_dim] < elements_[j][cur_dim]) {
        elements_[j + 1] = elements_[j];
        shifted = true;
        if (j == 0) {
          zero_reached = true;
          break;
        } else {
          --j;
        }
      }
      if (shifted) {
        if (zero_reached) {
          elements_[j] = key;
        } else {
          elements_[j + 1] = key;
        }
      }
    }
  }

  double CalculateVariation(int cur_dim, const size_t lower_idx, const size_t upper_idx) {
    if (!HasData() || lower_idx > upper_idx || upper_idx >= DataSize()) {
      return 0.0;
    }
    double mean = 0.0, var = 0.0;
    for (size_t i = lower_idx; i <= upper_idx; ++i) {
      mean += elements_[i][cur_dim];
    }
    mean /= static_cast<double>(DataSize());
    for (size_t i = lower_idx; i <= upper_idx; ++i) {
      double temp = elements_[i][cur_dim] - mean;
      var += temp * temp;
    }
    var /= static_cast<double>(DataSize());
    return var;
  }

  void PrintDataSet() {
    std::cout << "[PrintDataSet] " << std::endl;
    for (const auto& elem : elements_) {
      PrintElement(elem);
    }
    std::cout << std::endl;
  }

 private:
  std::vector<T> elements_;
  KDNodePtr root_ = nullptr;
  int dimension_ = 0;
  size_t k_ = 1;  // num of nearest points

  void PrintElement(const T& elem) {
    if (dimension_ <= 0) {
      return;
    }
    std::cout << "[";
    for (size_t i = 0; i < dimension_; ++i) {
      std::cout << elem[i];
      if (i < dimension_ - 1) {
        std::cout << ", ";
      }
    }
    std::cout << "]" << std::endl;
  }
};

}  // namespace lidar_localization

#endif

int main(int argc, char const *argv[])
{
  /* code */
  size_t a = 3ul;
  size_t b = 4ul;
  size_t c = (a+b)/2ul;
  std::cout << c << std::endl;

  // data_set = [[2, 3], [5, 4], [9, 6], [4, 7], [8, 1], [7, 2]]
  double data_set[6][2] = {2,3,5,4,9,6,4,7,8,1,7,2};
  lidar_localization::KDTree<Eigen::VectorXd> kd_tree;
  for (int i = 0; i < 6; ++i) {
    Eigen::VectorXd a(2);
    a(0) = data_set[i][0];
    a(1) = data_set[i][1];
    kd_tree.AddData(a,2);
  }
  kd_tree.PrintDataSet();
  kd_tree.GenerateTree();
  kd_tree.PrintTree(0);

  Eigen::VectorXd x(2);
  x(0) = 2.0;
  x(1) = 4.5;
  Eigen::VectorXd y(2);
  std::vector<Eigen::VectorXd> y_set;
  bool result = kd_tree.FindNearestPoint(x,y);
  result = kd_tree.FindNearestPoint(x,y_set,2);
  std::cout << "result: " << result << " nearest point: " << std::endl << y << std::endl;
  std::cout << "result: " << result << " nearest point: " << std::endl;
  for (auto y : y_set) {
    std::cout << y << std::endl;
  }

  return 0;
}

File: kd_tree.py
import math
import copy


class Queue():
    def __init__(self):
        self.q = list()

    def enq(self, v):
        self.q.append(v)

    def deq(self):
        if len(self.q) > 0:
            return self.q.pop(0)
        else:
            return None

    def empty(self):
        return len(self.q) == 0

    def __len__(self):
        return len(self.q)


class Stack():
    def __init__(self):
        self.s = list()

    def push(self, v):
        self.s.append(v)

    def pop(self):
        if len(self.s) > 0:
            return self.s.pop()
        else:
            return None

    def top(self):
        if len(self.s) > 0:
            return self.s[-1]
        else:
            return None

    def empty(self):
        return len(self.s) == 0

    def __len__(self):
        return len(self.s)


class Node:
    def __init__(self):
        self.range = None  # space range
        self.data = None  # k dimensional vector
        self.split = None  # the index of axis vertical to the split hyper surface
        self.left = None  # left child kd tree
        self.right = None  # right child kd tree
        self.parent = None  # parent node

    def SetParent(self, node):
        self.parent = node

    def SetLeftChild(self, node):
        self.left = node

    def SetRightChild(self, node):
        self.right = node

    def SetSplit(self, split):
        self.split = split

    def SetData(self, data):
        self.data = data

    def Data(self):
        return self.data

    def LeftChild(self):
        return self.left

    def RightChild(self):
        return self.right

    def Parent(self):
        return self.parent

    def HasLeftChild(self):
        return self.left != None

    def HasRightChild(self):
        return self.right != None

    def HasParent(self):
        return self.parent != None

    def Split(self):
        return self.split

    def HasSplit(self):
        return self.split != None

    def HasData(self):
        return self.data != None

    def IsLeftChild(self, node):
        if self.split == None or not node.HasData():
            return False
        return self.data[self.split] > node.Data()[self.split]

    def IsRightChild(self, node):
        if self.split == None or not node.HasData():
            return False
        return self.data[self.split] <= node.Data()[self.split]

    def IsLeaf(self):
        return not (self.HasLeftChild() or self.HasRightChild())


class KDTree:
    def __init__(self, data_set=None):
        self.root = Node()
        self.data_set = data_set
        self.dimension = None
        if data_set is not None and len(data_set) > 0:
            self.dimension = len(self.data_set[0])
            self.root = self.CreateNode(self.data_set)

    # maybe we have better method for dividing data
    def CalculateVariation(self, data_set, cur_dim):
        if len(data_set) == 0 or data_set is None:
            return 0
        mean = 0
        for data in data_set:
            mean += data[cur_dim]
        mean /= len(data_set)
        var = 0
        for data in data_set:
            var += math.pow(data[cur_dim]-mean, 2)
        var /= len(data_set)
        return var

    def InsertSort(self, data_set, cur_dim):
        if len(data_set) < 2:
            return
        for i in range(1, len(data_set)):
            key = copy.deepcopy(data_set[i])
            j = i-1
            while j >= 0 and data_set[j][cur_dim] > key[cur_dim]:
                data_set[j+1] = copy.deepcopy(data_set[j])
                j -= 1
            if j < i-1:
                data_set[j+1] = key

    def CreateNode(self, data_set, depth=0):
        node = Node()
        if len(data_set) == 0:
            print('data len=0, returning')
            return node
        if len(data_set) == 1:
            node.SetData(data_set[0])
            print('data: ', node.Data(), ' depth: ', depth)
            return node
        # determine split with largest variation
        split = 0
        max_var = 0
        for i in range(0, self.dimension):
            var = self.CalculateVariation(data_set, i)
            if var >= max_var:
                max_var = var
                split = i
        print('data len: ', len(data_set), 'split: ', split, ' depth: ', depth)
        node.SetSplit(split)
        # sort data set by split
        # and set mid value as node data
        self.InsertSort(data_set, split)
        mid_idx = int(len(data_set)/2)
        mid_data = data_set[mid_idx]
        node.SetData(mid_data)
        print('sorted set: ', data_set, 'mid idx: ',
              mid_idx, 'mid data', mid_data)
        # get left & right node
        depth += 1
        if mid_idx > 0:
            left_data_set = data_set[0:mid_idx]
            left_node = self.CreateNode(left_data_set, copy.deepcopy(depth))
            left_node.SetParent(node)
            node.SetLeftChild(left_node)
        if mid_idx+1 < len(data_set):
            right_data_set = data_set[mid_idx+1:len(data_set)]
            right_node = self.CreateNode(right_data_set, copy.deepcopy(depth))
            right_node.SetParent(node)
            node.SetRightChild(right_node)

        return node

    def PrintTree(self, print_type=1):
        print('printing tree ...')
        # use bfs/dfs to print tree
        if print_type == 1:
            q = Queue()
            q.enq(self.root)
            # print('level: ',step,' node: ',self.root.Data())
            while not q.empty():
                current_node = q.deq()
                print('current node: ', current_node.Data(),
                      ' split: ', current_node.Split())
                if current_node.HasLeftChild():
                    q.enq(current_node.LeftChild())
                if current_node.HasRightChild():
                    q.enq(current_node.RightChild())
        else:

            def dfs(node, s):
                s.push(node)
                step = 0
                if node.HasLeftChild():
                    dfs(node.LeftChild(), s)
                step += 1
                if node.HasRightChild():
                    dfs(node.RightChild(), s)
                step += 1
                # we need to print leaf node
                # and node that all its children has been printed
                if not(node.HasLeftChild() or node.HasRightChild()) or step == 2:
                    print('current node: ', node.Data())
                    s.pop()

            s = Stack()
            dfs(self.root, s)

        print('print tree done')

    def FindNearest(self, node, search_radius):
        def CalDist(n0, n1):  # L2 norm(maybe other norm possible?)
            sqr_sum = 0
            for i in range(0, len(n0.Data())):
                deviation = n0.Data()[i]-n1.Data()[i]
                sqr_sum += math.pow(deviation, 2)
            return math.sqrt(sqr_sum)

        def JudgeCutWithHyperSurface(node, target_node, min_dist):
            # judge a circle/ball/hypersphere ... with node as center, radius as radius
            # has intersection with hyper surface that is vertical to given axis
            # first project node onto surface
            radius = min(CalDist(node, target_node), min_dist)
            surface_axis = target_node.Split()
            surface_dist_to_origin = target_node.Data()[surface_axis]
            projection_node = Node()
            projection_node_data = copy.deepcopy(node.Data())
            projection_node_data[surface_axis] = surface_dist_to_origin
            projection_node.SetData(projection_node_data)
            # check if dist btwn two nodes is bigger than radius
            projection_dist = CalDist(node, projection_node)
            # method 2
            projection_dist = abs(
                node.Data()[surface_axis]-target_node.Data()[surface_axis])
            print('axis: ', surface_axis, ' dist to origin: ', surface_dist_to_origin,
                  ' surface dist:', projection_dist, ' radius: ', radius)
            if projection_dist > radius:
                return False
            return True

        # binary search & record search path
        current_node = self.root
        node_list = Stack()
        node_list.push(current_node)
        node_data_list = [current_node.Data()]
        print('current node has left: ', current_node.HasLeftChild())
        while (current_node.HasLeftChild() or current_node.HasRightChild()):
            print('current node: ', current_node.Data(),
                  ' split: ', current_node.Split())
            if not current_node.HasSplit():
                break
            current_split = current_node.Split()
            if current_node.Data()[current_split] > node.Data()[current_split]:
                current_node = current_node.LeftChild()
            else:
                current_node = current_node.RightChild()
            if CalDist(node, current_node) < search_radius:
                node_list.push(current_node)
                node_data_list.append(current_node.Data())
        print('node list: ', node_data_list)

        # reverse search to update possible closer node
        nearest_node = node_list.pop()
        min_dist = CalDist(node, nearest_node)

        print('begin reverse search, nearest:', nearest_node.Data())
        while not node_list.empty():
            print('node list len: ', len(node_list))
            target_node = node_list.pop()
            print('target node: ', target_node.Data(),
                  ' split: ', target_node.Split())
            # update nearest node
            temp_dist = CalDist(node, target_node)
            if min_dist > temp_dist:
                min_dist = temp_dist
                nearest_node = target_node
                print('update nearest: ', nearest_node.Data())
            # check the other branch
            if not target_node.IsLeaf():
                if not JudgeCutWithHyperSurface(node, target_node, min_dist):
                    # if node doesnt intersect with target node's hyper surface
                    # reverse search is done(if it does, there's a chance the other side
                    # has closer dist node)
                    print('no cut break')
                    break
                print('cut with surface')
                if target_node.IsLeftChild(node) and target_node.HasRightChild():
                    next_node = target_node.RightChild()
                    # push current node in node list
                    node_list.push(next_node)
                elif target_node.HasLeftChild():
                    next_node = target_node.LeftChild()
                    # push current node in node list
                    node_list.push(next_node)
                

        return nearest_node.Data()

    # Note this is NOT recommanded because
    # Adding points in this manner can cause the tree to become unbalanced, leading to decreased tree performance. The rate of tree performance degradation is dependent upon the spatial distribution of tree points being added, and the number of points added in relation to the tree size
    def AddNode(self, node):
         # binary search & record search path
        current_node = self.root
        node_list = Stack()
        node_list.push(current_node)
        node_data_list = [current_node.Data()]
        print('current node has left: ', current_node.HasLeftChild())
        while (current_node.HasLeftChild() or current_node.HasRightChild()):
            print('current node: ', current_node.Data(),
                  ' split: ', current_node.Split())
            if not current_node.HasSplit():
                break
            current_split = current_node.Split()
            if current_node.Data()[current_split] > node.Data()[current_split]:
                current_node = current_node.LeftChild()
            else:
                current_node = current_node.RightChild()
            node_list.push(current_node)
            node_data_list.append(current_node.Data())
        print('node list: ', node_data_list)

        if current_node.HasSplit():
            current_split = current_node.Split()
            if current_node.Data()[current_split] > node.Data()[current_split]:
                current_node.SetLeftChild(node)
            else:
                current_node.SetRightChild(node)
            node.SetParent(current_node)


data_set = [[2, 3], [5, 4], [9, 6], [4, 7], [8, 1], [7, 2]]
# data_set = [[19, 2], [7, 0], [13, 5], [3, 15], [3, 4], [3, 2], [8, 9], [9, 3], [17, 15], [11, 11]]
kd_tree = KDTree(data_set)
kd_tree.PrintTree()
target_node = Node()
# target_node.SetData([2.1,3.1])
target_node.SetData([2, 4.5])
# target_node.SetData([12, 7])
print('nearest node: ', kd_tree.FindNearest(target_node, 100))
